crates/turbopack-ecmascript/src/swc_comments.rs

```diff
@@ -0,0 +1,122 @@
+use std::{collections::HashMap, mem::take};
+
+use swc_core::{
+    base::SwcComments,
+    common::{
+        comments::{Comment, Comments},
+        BytePos,
+    },
+};
+
+/// Immutable version of [SwcComments] which doesn't allow mutation. The `take`
+/// variants are still implemented, but do not mutate the content. They are used
+/// by the SWC Emitter.
+pub struct ImmutableComments {
+    pub leading: HashMap<BytePos, Vec<Comment>>,
+    pub trailing: HashMap<BytePos, Vec<Comment>>,
+}
+
+impl ImmutableComments {
+    pub fn new(comments: SwcComments) -> Self {
+        Self {
+            leading: comments
+                .leading
+                .iter_mut()
+                .filter_map(|mut r| {
+                    let c = take(r.value_mut());
+                    (!c.is_empty()).then_some((*r.key(), c))
+                })
+                .collect(),
+            trailing: comments
+                .trailing
+                .iter_mut()
+                .filter_map(|mut r| {
+                    let c = take(r.value_mut());
+                    (!c.is_empty()).then_some((*r.key(), c))
+                })
+                .collect(),
+        }
+    }
+}
+
+impl Comments for ImmutableComments {
+    fn add_leading(
+        &self,
+        _pos: swc_core::common::BytePos,
+        _cmt: swc_core::common::comments::Comment,
+    ) {
+        panic!("Comments are immutable after parsing")
+    }
+
+    fn add_leading_comments(
+        &self,
+        _pos: swc_core::common::BytePos,
+        _comments: Vec<swc_core::common::comments::Comment>,
+    ) {
+        panic!("Comments are immutable after parsing")
+    }
+
+    fn has_leading(&self, pos: swc_core::common::BytePos) -> bool {
+        self.leading.contains_key(&pos)
+    }
+
+    fn move_leading(&self, _from: swc_core::common::BytePos, _to: swc_core::common::BytePos) {
+        panic!("Comments are immutable after parsing")
+    }
+
+    fn take_leading(
+        &self,
+        pos: swc_core::common::BytePos,
+    ) -> Option<Vec<swc_core::common::comments::Comment>> {
+        self.get_leading(pos)
+    }
```

Comment:
Shoudn't this throw, since the name sounds like it's taking the comment at that position?

Reply:
yeah, sadly the `Emitter` calls this method on comment. I guess to avoid cloning the comment. So we need to implement it.

Technically cloning an `Comment` isn't that expensive (Copy 13 bytes and Clone and Arc), so I would prefer the `Emitter` using `get_leading` instead of `take_leading`, but it is as it is.

There is a comment on the struct explaining the `take` behavior

Reply:
TBH it's not very obvious that the `Emitter` mutates the passed comments. This is how this bug sneaked into turbopack @kdy1